### PR TITLE
fix: runtime select scripting or content scripts apis

### PR DESCRIPTION
**Pull request by rpl**
This PR is a followup to the previous set of changes introduced in #320, and part of handling the issue caught when we have been vendored the new webcompat version using the scripting API in mozilla-central ([Bug 1853013](https://bugzilla.mozilla.org/show_bug.cgi?id=1853013)).

The underlying issue is due to the scripting.registerContentScripts API currently forcefully setting matchAboutBlank internally to `true` ([as also described in Bug 1853013 comment 6](https://bugzilla.mozilla.org/show_bug.cgi?id=1853013#c6).

The approach we discussed and agreed on with @wisniewskit earlier today is composed of two parts:

- one part of the approach is this set of changes on the webcompat side, this changes are meant to let webcompat to select either scripting or contentScripts API at runtime (and so this PR is largely about bringing back the support for contentScripts.register and select one of the API based on a call to a new experimental API, which will be synchronously provide a boolean result to determine if scripting should be used in the current browser instance)

- the second part of the issue will have to apply some change on the `scripting` API implementation on the mozilla-central side, tracked by [Bug 1853412 - Set matchAboutBlank to false in scripting.registerContentScripts for webcompat](https://bugzilla.mozilla.org/show_bug.cgi?id=1853412)

In this first version of the patch I've brought back the code for using contentScripts.register by also trying to contain the changes in the same area that we have touched in #320 without a too big of a refactoring, there may be some more tweaks we can still apply without expanding the impact of the changes but we may also likely want to balance between what we could consider reasonable in this PR vs. in separate followups and so this version felt like a reasonable starting point.

## TODO list

- [x] roll the new experimental API method into `aboutConfigPrefs` (as a new `getBoolPrefSync` method running in the child extension process)
- [x] add at least one additional small test to cover Injections falling back to contentScripts.register based on the value got for the useScriptingAPI pref
- [x] add at least a couple small tests to cover Shims registering content scripts with scripting and contentScripts API based on the value got for the useScriptingAPI pref
- [x] link bugzilla issue tracking the changes on the scripting API implementation side in mozilla-central
- [x] ~~apply changes to add matchAboutBlank option if that is going to be the option we will choose for the changes applied on the scripting API implementation side~~ (dropped as not actually needed, the option we are going for in the short term in the one tracked by [Bug 1853412 - Set matchAboutBlank to false in scripting.registerContentScripts for webcompat](https://bugzilla.mozilla.org/show_bug.cgi?id=1853412) and that doesn't need further changes to this PR.

### CODE REVIEW

**Comment by rpl** [comment]
@wisniewskit this is the initial draft for the PR we discussed earlier today, in the PR description I've added some more details about the approach (in particular the new experimental API checking the pref and synchronously providing a boolean result for which API to pick).

I've marked it as a draft while I'm working with my team to determine which option for the changes on the scripting API implementation side we are going to take, but in the meantime you may take a look to this version

Also, thanks again for the help!!! :heart:

**Comment by rpl** [comment]
Push to try including this PR + one line change to default matchAboutBlank to false for webcompat in the short term (which is one of the options being considered, in alternative to explicitly introducing matchAboutBlank option in the scripting.registerContentScripts API method) + flipping the `extensions.webcompat.useScriptingAPI` pref to `true` to opt-in on using the scripting API:

- https://treeherder.mozilla.org/jobs?repo=try&revision=95ee7165089ce3a87c8a05c6294edda2be7d2517

**Comment by rpl** [comment]
Given that the new experimental API was mainly synchronously returning the value of a pref in the form it was when I created this PR, and both @wisniewskit and @willdurand were wondering if we could be rolling the new API method into one of the existing APIs, I decided to go on and roll that method into the aboutConfigPrefs experimental API (which is now including a separate aboutConfigPrefsChild.js which is the part of the API running in the child extension process and providing the getBoolPrefSync method which is meant to return the value of a webcompat pref synchrously (which for consistency with the getPref method is implicitly prefixing the pref name with `extensions.webcompat`).

Besides that tweak, the updated version of this PR is functionally the same as the version submitted yesterday.

**Comment by rpl** [comment]
@wisniewskit this PR is now officially open for review, the change on the scripting API internals side is tracked by [Bug 1853412 - Set matchAboutBlank to false in scripting.registerContentScripts for webcompat](https://bugzilla.mozilla.org/show_bug.cgi?id=1853412).

[Bug 1853412](https://bugzilla.mozilla.org/show_bug.cgi?id=1853412) patches are including adding the "extensions.webcompat.useScriptingAPI" pref set to true, because I thought that may better fit into a separate patch landed along with [Bug 1853013](https://bugzilla.mozilla.org/show_bug.cgi?id=1853013), does that sound good to you?

The pref should be set to true in ` modules/libpref/init/all.js` to make sure it is also set to true for GeckoView builds (so that once we will be vendoring the updated webcompat in firefox-android repo then the GeckoView builds will also be opting-in on using the scripting API.

If you add me as a reviewer on the patch that sets the pref I'll be happy to sign it off.

As a side note, today I applied the following changes:
- merge the new API method into the aboutConfigPref experimental API and renamed it to just getBoolPrefSync
- added some more coverage for the fallback mechanism for both Injections and Shims.

Besides these couple of changes the PR is the same as it was in its initial version yesterday.

**Comment by rpl** [comment]
@wisniewskit The last change applied today is purely cosmetic, nothing changes in practice besides being more immediately visible that the second parameter is optional and defaults to false.

I've also pushed to try a couple of patches (vendoring these changes in m-c + opt-in on using the scripting API) and linked it here in [Bug 1853013 comment 8](https://bugzilla.mozilla.org/show_bug.cgi?id=1853013#c8).

File: spec/injections.spec.js [hunk 1]
```diff
@@ -124,6 +95,22 @@ describe("Injections", () => {
...
+      // getBoolPrefSync("useScriptingAPI") returns false.
+      let spyGetBookPrefSync = spyOn(
+        browser.aboutConfigPrefs,
+        "getBoolPrefSync"
+      ).and.returnValue(false);
+
+      let injections = new Injections([buildInjection("desktop")]);
+      expect(spyGetBookPrefSync).toHaveBeenCalledOnceWith("useScriptingAPI");
+      injections.bindAboutCompatBroker(mockBroker);
+      let spyRegister = spyOn(browser.contentScripts, "register");
```
Comment (Rob--W) [inline spec/injections.spec.js:110]: Should we also test that there are no scripting API calls when contentScripts are used, and the other way around? In theory there may be duplicate registrations.

**Review by Rob--W — APPROVED** [review]
I confirmed that scripting / contentScripts are mutually exclusive here. Looks good to merge from my POV.

File: src/experiment-apis/aboutConfigPrefsChild.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    const extensionIDBase = context.extension.id.split("@")[0];
+    const extensionPrefNameBase = `extensions.${extensionIDBase}.`;
+
+    return {
+      aboutConfigPrefs: {
+        getBoolPrefSync(prefName, defaultValue = false) {
+          try {
+            return Services.prefs.getBoolPref(
+              `${extensionPrefNameBase}${prefName}`,
+              defaultValue
```
[the remark this answers is not in the document]
Comment (rpl) [reply src/experiment-apis/aboutConfigPrefsChild.js:20]: under normal conditions there shouldn't be, but to be fair from a quick look to the internal C++ implementation it seems there are technically other kind of unexpected conditions that could lead to an error being raised (e.g. if `Preferences::InitStaticMembers()` failed, and if I'm reading it correctly also if there is an unexpected mismatch between the value actually set and the one we expect, boolean)

And given this is a more generic method that could be used in the future not just for the check of the `extensions.webcompat.useScriptingAPI` pref, I thought to keep the exact same convention used by the other similar method defined in aboutConfigPrefs.

personally I prefer the two method to look the same, and if changed then both to be changed to stay as consistent with each other as possible, and so I don't see that as strictly something to be done in this particular PR.

File: src/experiment-apis/aboutConfigPrefsChild.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+            );
+          } catch (_) {
```
[the remark this answers is not in the document]
Comment (rpl) [reply src/experiment-apis/aboutConfigPrefsChild.js:22]: I'd leave it as is given that the other getPref method is written like this, they can eventually be both changed in a separate followup, in this PR I'd prefer to keep the two method as in sync as possible (as also mentioned in my answer to another review comment).

File: src/experiment-apis/aboutConfigPrefsChild.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+"use strict";
+
+/* global ExtensionAPI, Services, XPCOMUtils */
```
[the remark this answers is not in the document]
Comment (rpl) [reply src/experiment-apis/aboutConfigPrefsChild.js:7]: :+1: sure, I can remove the unnecessary one from this eslint config inline comment.

File: spec/injections.spec.js (hunk 1 above)
Comment (rpl) [reply spec/injections.spec.js:110]: :+1: That sounds like a good idea, I'm looking into adding that to this test case.

Comment (rpl) [reply spec/injections.spec.js:118]: I added the additional assertion in [c7a57dd](https://github.com/mozilla-extensions/webcompat-addon/pull/324/commits/c7a57ddad8410aaa1b50a7ccfea78ec7769c727d).

File: src/experiment-apis/aboutConfigPrefsChild.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (rpl) [reply src/experiment-apis/aboutConfigPrefsChild.js:7]: Removed in [899cde4](https://github.com/mozilla-extensions/webcompat-addon/pull/324/commits/899cde4dc8c9b273fe4d57825cc035ec6e37ca06).

**Review by wisniewskit — APPROVED** [review]
Ok, looks good to me. Let's try to land this right away in case we find anything strange again before the soft-freeze.

**Pull request merged**